File: Miner.go
package main

import (
	"bytes"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/gob"
	"encoding/hex"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"io/ioutil"
	"log"
	"math"
	"math/big"
	"os"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

type Block struct {
	Index        int
	Timestamp    string
	Transactions []Transaction
	PreviousHash string
	Hash         string
	Nonce        int
	Miner        string
	Reward       float64
}

type Transaction struct {
	From      string  `json:"from"`
	To        string  `json:"to"`
	Amount    float64 `json:"amount"`
	Timestamp string  `json:"timestamp"`
	Signature string  `json:"signature"`
	PubKey    string  `json:"pubKey"` // public key pengirim
}

type Blockchain struct {
	Chain      []Block
	Difficulty int
	sync.Mutex
}

type Member struct {
	Username string  `json:"username"`
	Address  string  `json:"address"`
	Balance  float64 `json:"balance"`
	Joined   string  `json:"joined"`
}

var blockchain = &Blockchain{
	Difficulty: 5,
}

const blockFolder = "./blockchain_files"
const halvingFile = "halving.zthx"
const halvingInterval = 10

func generateKeyPair() (*ecdsa.PrivateKey, *ecdsa.PublicKey) {
	priv, _ := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	return priv, &priv.PublicKey
}

func savePrivateKey(priv *ecdsa.PrivateKey, filename string) {
	privBytes, _ := x509.MarshalECPrivateKey(priv)
	pemBlock := &pem.Block{Type: "EC PRIVATE KEY", Bytes: privBytes}
	pemData := pem.EncodeToMemory(pemBlock)
	_ = os.WriteFile(filename, pemData, 0600)
}

func savePublicKey(pub *ecdsa.PublicKey, filename string) {
	pubBytes, _ := x509.MarshalPKIXPublicKey(pub)
	pemBlock := &pem.Block{Type: "PUBLIC KEY", Bytes: pubBytes}
	pemData := pem.EncodeToMemory(pemBlock)
	_ = os.WriteFile(filename, pemData, 0644)
}

func publicKeyToAddress(pub *ecdsa.PublicKey) string {
	pubBytes, _ := x509.MarshalPKIXPublicKey(pub)
	hash := sha256.Sum256(pubBytes)
	return "zyth-" + hex.EncodeToString(hash[:8])
}

func signTransaction(priv *ecdsa.PrivateKey, tx Transaction) string {
	data := tx.From + tx.To + fmt.Sprintf("%f", tx.Amount) + tx.Timestamp
	hash := sha256.Sum256([]byte(data))
	r, s, _ := ecdsa.Sign(rand.Reader, priv, hash[:])
	sig := append(r.Bytes(), s.Bytes()...)
	return hex.EncodeToString(sig)
}

func verifyTransaction(tx Transaction) bool {
	// Coinbase transaction tidak perlu verifikasi
	if tx.From == "SYSTEM" {
		return true
	}

	data := tx.From + tx.To + fmt.Sprintf("%f", tx.Amount) + tx.Timestamp
	hash := sha256.Sum256([]byte(data))

	sigBytes, err := hex.DecodeString(tx.Signature)
	if err != nil {
		return false
	}
	r := new(big.Int).SetBytes(sigBytes[:len(sigBytes)/2])
	s := new(big.Int).SetBytes(sigBytes[len(sigBytes)/2:])

	block, _ := pem.Decode([]byte(tx.PubKey))
	if block == nil {
		return false
	}
	pubKeyIfc, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return false
	}
	pubKey := pubKeyIfc.(*ecdsa.PublicKey)

	return ecdsa.Verify(pubKey, hash[:], r, s)
}

func calculateHash(block Block) string {
	record := strconv.Itoa(block.Index) + block.Timestamp + fmt.Sprintf("%v", block.Transactions) + block.PreviousHash + strconv.Itoa(block.Nonce)
	h := sha256.New()
	h.Write([]byte(record))
	return fmt.Sprintf("%x", h.Sum(nil))
}

func mineBlock(block *Block, difficulty int) {
	prefix := strings.Repeat("0", difficulty)
	for {
		hash := calculateHash(*block)
		if strings.HasPrefix(hash, prefix) {
			block.Hash = hash
			break
		}
		block.Nonce++
		if block.Nonce%100000 == 0 {
			fmt.Printf("Mining... index: %d, nonce: %d, hash: %s\n", block.Index, block.Nonce, hash)
		}
	}
}

func createTransaction(from, to string, amount float64) Transaction {
	return Transaction{
		From:      from,
		To:        to,
		Amount:    amount,
		Timestamp: time.Now().Format(time.RFC3339),
	}
}

const mempoolFile = "pending_transactions.json"

func addTransaction(tx Transaction) {
	var txs []Transaction
	data, err := ioutil.ReadFile(mempoolFile)
	if err == nil {
		_ = json.Unmarshal(data, &txs)
	}
	txs = append(txs, tx)
	jsonData, _ := json.MarshalIndent(txs, "", "  ")
	_ = ioutil.WriteFile(mempoolFile, jsonData, 0644)
}

func getPendingTransactions() []Transaction {
	var txs []Transaction
	data, err := ioutil.ReadFile(mempoolFile)
	if err == nil {
		_ = json.Unmarshal(data, &txs)
	}
	return txs
}

func clearMempool() {
	_ = ioutil.WriteFile(mempoolFile, []byte("[]"), 0644)
}

func getBalance(address string) float64 {
	balance := 0.0
	for _, block := range blockchain.Chain {
		for _, tx := range block.Transactions {
			if tx.From == address {
				balance -= tx.Amount
			}
			if tx.To == address {
				balance += tx.Amount
			}
		}
	}
	return balance
}

func validateTransaction(tx Transaction) bool {
	if !verifyTransaction(tx) {
		fmt.Println("❌ Invalid signature")
		return false
	}
	if tx.From != "SYSTEM" {
		balance := getBalance(tx.From)
		if balance < tx.Amount {
			fmt.Printf("❌ Insufficient balance for %s\n", tx.From)
			return false
		}
	}
	return true
}

func saveBlockAsZythFile(block Block) {
	os.MkdirAll(blockFolder, os.ModePerm)
	raw := fmt.Sprintf("block_%d", block.Index)
	hash := sha256.Sum256([]byte(raw))
	encoded := base64.URLEncoding.EncodeToString(hash[:])
	filename := fmt.Sprintf("%s/%s.zyth", blockFolder, encoded[:20])

	// Save binary (.zyth)
	var buf bytes.Buffer
	enc := gob.NewEncoder(&buf)
	err := enc.Encode(block)
	if err == nil {
		_ = ioutil.WriteFile(filename, buf.Bytes(), 0644)
	}

	// Save JSON for web access
	jsonPath := fmt.Sprintf("%s/%s.json", blockFolder, encoded[:20])
	jsonData, _ := json.MarshalIndent(block, "", "  ")
	_ = ioutil.WriteFile(jsonPath, jsonData, 0644)

	fmt.Printf("✔ Block %d saved to %s\n", block.Index, filename)
}

func createGenesisBlock() Block {
	fmt.Println("No chain found. Creating genesis block...")
	genesis := Block{
		Index:     0,
		Timestamp: time.Now().Format(time.RFC3339),
		Transactions: []Transaction{
			{From: "SYSTEM", To: "GENESIS", Amount: 0, Timestamp: time.Now().Format(time.RFC3339)},
		},
		PreviousHash: "0",
		Nonce:        0,
	}
	mineBlock(&genesis, blockchain.Difficulty)
	genesis.Hash = calculateHash(genesis)
	saveBlockAsZythFile(genesis)
	return genesis
}

func loadAllBlocks() {
	files, err := ioutil.ReadDir(blockFolder)
	if err != nil {
		blockchain.Chain = append(blockchain.Chain, createGenesisBlock())
		return
	}

	var loaded []Block
	for _, f := range files {
		if strings.HasSuffix(f.Name(), ".zyth") {
			data, err := ioutil.ReadFile(fmt.Sprintf("%s/%s", blockFolder, f.Name()))
			if err != nil {
				continue
			}
			var blk Block
			err = gob.NewDecoder(bytes.NewReader(data)).Decode(&blk)
			if err != nil {
				continue
			}
			loaded = append(loaded, blk)
		}
	}

	if len(loaded) == 0 {
		blockchain.Chain = append(blockchain.Chain, createGenesisBlock())
		return
	}

	sort.Slice(loaded, func(i, j int) bool {
		return loaded[i].Index < loaded[j].Index
	})
	blockchain.Chain = loaded
	fmt.Printf("Loaded %d blocks from file.\n", len(loaded))
}

func getReward(index int) float64 {
	raw, err := ioutil.ReadFile(halvingFile)
	if err != nil {
		log.Fatalf("Cannot read halving file: %v", err)
	}
	base, err := strconv.ParseFloat(strings.TrimSpace(string(raw)), 64)
	if err != nil {
		log.Fatalf("Invalid halving base: %v", err)
	}
	halvings := index / halvingInterval
	reward := base / math.Pow(2, float64(halvings))
	return reward
}

func mineSingle(address string) {
	loadAllBlocks()
	prev := blockchain.Chain[len(blockchain.Chain)-1]

	// Ambil transaksi pending
	pending := getPendingTransactions()

	// Filter transaksi valid
	var validTxs []Transaction
	for _, tx := range pending {
		if validateTransaction(tx) {
			validTxs = append(validTxs, tx)
		} else {
			fmt.Printf("❌ Invalid transaction: %+v\n", tx)
		}
	}

	// Tambahkan coinbase transaction (reward ke miner)
	reward := getReward(len(blockchain.Chain))
	coinbase := createTransaction("SYSTEM", address, reward)

	// Gabungkan coinbase + transaksi valid
	txs := append([]Transaction{coinbase}, validTxs...)

	newBlock := Block{
		Index:        len(blockchain.Chain),
		Timestamp:    time.Now().Format(time.RFC3339),
		Transactions: txs,
		PreviousHash: prev.Hash,
		Nonce:        0,
		Miner:        address,
		Reward:       reward,
	}

	fmt.Printf("Mining block %d by %s...\n", newBlock.Index, address)
	mineBlock(&newBlock, blockchain.Difficulty)
	newBlock.Hash = calculateHash(newBlock)

	blockchain.Lock()
	blockchain.Chain = append(blockchain.Chain, newBlock)
	blockchain.Unlock()

	saveBlockAsZythFile(newBlock)

	// Kosongkan mempool
	clearMempool()

	fmt.Printf("✅ Block %d mined! Hash: %s | Reward: %.8f ZYTH\n",
		newBlock.Index, newBlock.Hash, reward)
}

func mineGroup(address, roomCode string) {
	loadAllBlocks()
	prev := blockchain.Chain[len(blockchain.Chain)-1]

	// Baca member dari file room
	roomFile := fmt.Sprintf("room_data.json_%s", roomCode)
	roomBytes, err := ioutil.ReadFile(roomFile)
	if err != nil {
		log.Fatalf("❌ Failed to read room file: %v", err)
	}

	var room struct {
		Members []Member `json:"members"`
	}

	if err := json.Unmarshal(roomBytes, &room); err != nil {
		log.Fatalf("❌ Failed to parse room file: %v", err)
	}

	rewardTotal := getReward(len(blockchain.Chain))
	share := rewardTotal / float64(len(room.Members))

	rewardMap := make(map[string]float64)
	for _, m := range room.Members {
		rewardMap[m.Address] = share
	}

	var txs []Transaction
	for _, m := range room.Members {
		txs = append(txs, createTransaction("SYSTEM", m.Address, share))
	}

	newBlock := Block{
		Index:        len(blockchain.Chain),
		Timestamp:    time.Now().Format(time.RFC3339),
		Transactions: txs,
		PreviousHash: prev.Hash,
		Nonce:        0,
		Miner:        address,
		Reward:       rewardTotal,
	}

	fmt.Printf("Mining block %d by %s...\n", newBlock.Index, address)
	mineBlock(&newBlock, blockchain.Difficulty)
	newBlock.Hash = calculateHash(newBlock)

	blockchain.Lock()
	blockchain.Chain = append(blockchain.Chain, newBlock)
	blockchain.Unlock()

	saveBlockAsZythFile(newBlock)

	fmt.Printf("✅ Block %d mined! Hash: %s | Total Reward: %.8f ZYTH → %.8f each\n",
		newBlock.Index, newBlock.Hash, rewardTotal, share)
}

func main() {
	if len(os.Args) < 3 {
		fmt.Println("Usage: go run miner.go <your_address> <room_code>")
		return
	}
	address := os.Args[1]
	roomCode := os.Args[2]

	fmt.Printf(" Starting continuous mining for %s (room %s)...\n", address, roomCode)
	for {
		mineGroup(address, roomCode)
	}
}

File: miner.py
import json, os, time, hashlib, base64, glob
from datetime import datetime
import threading
import requests, json

# ANSI Escape Codes for coloring the output
# We'll use a combination of purple and white, with bold text for emphasis.
class Color:
    RESET = "\033[0m"
    BOLD = "\033[1m"
    WHITE = "\033[97m"
    PURPLE = "\033[95m"
    # A simple utility for a solid white block in text charts
    BLOCK_WHITE = "\033[47m \033[0m" # white background
    BLOCK_PURPLE = "\033[45m \033[0m" # purple background

BLOCK_FOLDER = "./blockchain_files"
MEMPOOL_FILE = "pending_transactions.json"
HALVING_FILE = "halving.zthx"
HALVING_INTERVAL = 10
DIFFICULTY = 5

# ---------------- Block & Tx structure ----------------
class Transaction:
    def __init__(self, from_addr, to_addr, amount, timestamp=None):
        self.from_addr = from_addr
        self.to_addr = to_addr
        self.amount = amount
        self.timestamp = timestamp or datetime.utcnow().isoformat()

    def to_dict(self):
        return {
            "from": self.from_addr,
            "to": self.to_addr,
            "amount": self.amount,
            "timestamp": self.timestamp
        }

class Block:
    def __init__(self, index, prev_hash, transactions, miner, reward):
        self.index = index
        self.timestamp = datetime.utcnow().isoformat()
        self.transactions = transactions
        self.previous_hash = prev_hash
        self.nonce = 0
        self.miner = miner
        self.reward = reward
        self.hash = ""

    def calculate_hash(self):
        record = f"{self.index}{self.timestamp}{self.transactions}{self.previous_hash}{self.nonce}"
        return hashlib.sha256(record.encode()).hexdigest()

# ---------------- Blockchain Utils ----------------
def load_chain():
    if not os.path.exists(BLOCK_FOLDER):
        os.makedirs(BLOCK_FOLDER)
        return [create_genesis_block()]
    files = glob.glob(os.path.join(BLOCK_FOLDER, "*.zthx"))
    if not files:
        return [create_genesis_block()]
    blocks = []
    for f in files:
        with open(f, "r") as fp:
            try:
                blk = json.load(fp)
                blocks.append(blk)
            except:
                continue
    blocks.sort(key=lambda b: b["Index"])
    return blocks

def create_genesis_block():
    print(f"\n{Color.BOLD}{Color.PURPLE}⚡ Creating Genesis Block...{Color.RESET}")
    genesis = {
        "Index": 0,
        "Timestamp": datetime.utcnow().isoformat(),
        "Transactions": [{"from": "SYSTEM", "to": "GENESIS", "amount": 0, "timestamp": datetime.utcnow().isoformat()}],
        "PreviousHash": "0",
        "Nonce": 0,
        "Miner": "GENESIS",
        "Reward": 0,
        "Hash": ""
    }
    genesis["Hash"] = hashlib.sha256(str(genesis).encode()).hexdigest()
    save_block(genesis)
    return genesis

def save_block(block):
    raw = f"block_{block['Index']}"
    encoded = base64.urlsafe_b64encode(hashlib.sha256(raw.encode()).digest()).decode()
    file_path = os.path.join(BLOCK_FOLDER, encoded[:20] + ".zthx")
    with open(file_path, "w") as f:
        json.dump(block, f, indent=2)
    print(f"{Color.WHITE}✔ Block {block['Index']} saved → {file_path}{Color.RESET}")

def get_reward(index):
    if not os.path.exists(HALVING_FILE):
        open(HALVING_FILE, "w").write("50")  # default reward
    base = float(open(HALVING_FILE).read().strip())
    halvings = index // HALVING_INTERVAL
    return base / (2 ** halvings)

def mine_block(block, difficulty):
    prefix = "0" * difficulty
    while True:
        h = block.calculate_hash()
        if h.startswith(prefix):
            block.hash = h
            return
        block.nonce += 1
        if block.nonce % 100000 == 0:
            print(f"{Color.PURPLE} Mining block : {block.index}, nonce : {block.nonce}, hash : {h}...{Color.RESET}")

PENDING_FILE = "pending_transactions.json"

def get_pending_transactions():
    if not os.path.exists(PENDING_FILE):
        return []
    try:
        return json.load(open(PENDING_FILE,"r"))
    except:
        return []

def broadcast_block(block_data):
    try:
        peers = json.load(open("p2p/peers.json"))
    except:
        peers = []

    for peer in peers:
        try:
            res = requests.post(f"http://{peer}/new_block", json=block_data, timeout=3)
            print(f"Broadcast to {peer} → {res.status_code}")
        except Exception as e:
            print(f"Failed to reach {peer}: {e}")
            
def clear_mempool():
    json.dump([], open(PENDING_FILE,"w"))

def print_banner():
    """Prints a large, stylized banner for the CLI."""
    print(f"{Color.BOLD}{Color.PURPLE}")
    print(f"{Color.BOLD}{Color.WHITE}     - ZYTHERION MINING CLI -{Color.RESET}\n")

def print_chart(chain):
    """Prints a simple text-based chart showing block progress."""
    index = len(chain)
    blocks_to_halving = HALVING_INTERVAL - (index % HALVING_INTERVAL)
    progress_percent = int(((HALVING_INTERVAL - blocks_to_halving) / HALVING_INTERVAL) * 100)
    
    # Create the colored progress bar
    bar_length = 30
    filled_length = int(bar_length * progress_percent / 100)
    bar = f"{Color.PURPLE}{'█' * filled_length}{Color.WHITE}{'░' * (bar_length - filled_length)}{Color.RESET}"
    
    print(f"{Color.BOLD}{Color.WHITE}--- Blockchain Status ---{Color.RESET}")
    print(f"Block Count: {Color.PURPLE}{index}{Color.RESET}")
    print(f"Next Halving in: {Color.PURPLE}{blocks_to_halving}{Color.RESET} blocks")
    print(f"Progress: {bar} {progress_percent}%")
    print("-" * 25 + "\n")

# ---------------- Mining ----------------
def mine_single(address):
    chain = load_chain()
    prev = chain[-1]

    # 🔹 ambil pending transaction
    txs = get_pending_transactions()

    # 🔹 tambahkan coinbase reward (hadiah mining)
    reward = get_reward(len(chain))
    coinbase = Transaction("SYSTEM", address, reward).to_dict()
    txs.insert(0, coinbase)

    new_block = Block(len(chain), prev["Hash"], txs, address, reward)
    print(f"\n{Color.PURPLE} Mining block {new_block.index} solo for {address}...{Color.RESET}")
    mine_block(new_block, DIFFICULTY)

    block_data = {
        "Index": new_block.index,
        "Timestamp": new_block.timestamp,
        "Transactions": new_block.transactions,
        "PreviousHash": prev["Hash"],
        "Nonce": new_block.nonce,
        "Miner": address,
        "Reward": reward,
        "Hash": new_block.hash
    }
    save_block(block_data)
    broadcast_block(block_data)
    # 🔹 kosongkan mempool
    clear_mempool()

    print(f"\n{Color.BOLD}{Color.WHITE} Solo block {new_block.index} mined! Reward {reward:.2f} ZYTH{Color.RESET}")
    print_chart(load_chain())


def mine_group(address, room_code):
    chain = load_chain()
    prev = chain[-1]

    room_file = f"room_data.json_{room_code}"
    if not os.path.exists(room_file):
        print(f"{Color.PURPLE} Room file not found{Color.RESET}")
        return

    room = json.load(open(room_file))
    members = room.get("members", [])
    if not members:
        print(f"{Color.PURPLE} No members in room{Color.RESET}")
        return

    reward_total = get_reward(len(chain))
    share = reward_total / len(members)

    # 🔹 ambil pending tx dulu
    txs = get_pending_transactions()

    # 🔹 bagi reward ke semua anggota
    for m in members:
        txs.append(Transaction("SYSTEM", m["address"], share).to_dict())

    new_block = Block(len(chain), prev["Hash"], txs, address, reward_total)
    print(f"\n{Color.PURPLE}⛏ Group mining block {new_block.index} for room {room_code}...{Color.RESET}")
    mine_block(new_block, DIFFICULTY)

    block_data = {
        "Index": new_block.index,
        "Timestamp": new_block.timestamp,
        "Transactions": new_block.transactions,
        "PreviousHash": prev["Hash"],
        "Nonce": new_block.nonce,
        "Miner": address,
        "Reward": reward_total,
        "Hash": new_block.hash
    }
    save_block(block_data)
    broadcast_block(block_data)
    # 🔹 kosongkan mempool
    clear_mempool()

    print(f"\n{Color.BOLD}{Color.WHITE} Group block {new_block.index} mined! Total {reward_total:.2f} ZYTH → {share:.2f} each{Color.RESET}")
    print_chart(load_chain())


# ---------------- CLI ----------------
if __name__ == "__main__":
    import sys

    print_banner()
    
    if len(sys.argv) < 3:
        print(f"{Color.BOLD}{Color.PURPLE}Usage:{Color.RESET}")
        print(f"{Color.WHITE} python miner.py solo {Color.PURPLE}<your_address>{Color.RESET}")
        print(f"{Color.WHITE} python miner.py group {Color.PURPLE}<your_address> <room_code>{Color.RESET}")
        sys.exit(1)
        
    print_chart(load_chain())

    mode = sys.argv[1]
    if mode == "solo":
        address = sys.argv[2]
        while True:
            mine_single(address)
    elif mode == "group":
        if len(sys.argv) < 4:
            print(f"{Color.PURPLE}Need <your_address> and <room_code>{Color.RESET}")
            sys.exit(1)
        address = sys.argv[2]
        room_code = sys.argv[3]
        while True:
            mine_group(address, room_code)
